diff --git a/drivers/mtd/tplinkpart.c b/drivers/mtd/tplinkpart.c
index 367b161..e460c61 100644
--- a/drivers/mtd/tplinkpart.c
+++ b/drivers/mtd/tplinkpart.c
@@ -16,7 +16,7 @@
 #include <linux/mtd/mtd.h>
 #include <linux/mtd/partitions.h>
 
-#define TPLINK_NUM_PARTS	5
+#define TPLINK_NUM_PARTS	6
 #define TPLINK_HEADER_V1	0x01000000
 #define MD5SUM_LEN		16
 
@@ -158,8 +158,9 @@ static int tplink_parse_partitions(struct mtd_info *master,
 	parts[2].name = "rootfs";
 	parts[2].offset = rootfs_offset;
 	parts[2].size = art_offset - rootfs_offset;
+
 #define FACTORY_LAYOUT 1
-#ifdef FACTORY_LAYOUT
+#ifdef FACTORY_LAYOUT_ORIG
 	parts[3].name = "firmware";
 	parts[3].offset = offset;
 	parts[3].size = art_offset - offset;
@@ -168,6 +169,35 @@ static int tplink_parse_partitions(struct mtd_info *master,
 	parts[4].offset = art_offset;
 	parts[4].size = TPLINK_ART_LEN;
 /*	parts[4].mask_flags = MTD_WRITEABLE; */
+#elif defined(FACTORY_LAYOUT)
+/*
+	parts[2].size = 0x800000 - rootfs_offset;
+
+	parts[3].name = "rootfs_data";
+	parts[3].offset = 0x800000;
+	parts[3].size = art_offset - 0x800000;
+
+	parts[4].name = "firmware";
+	parts[4].offset = offset;
+	parts[4].size = art_offset - 0x800000;
+
+	parts[5].name = "art";
+	parts[5].offset = art_offset;
+	parts[5].size = TPLINK_ART_LEN;
+*/
+
+	parts[3].name = "firmware";
+	parts[3].offset = offset;
+	parts[3].size = art_offset - offset;
+
+	parts[4].name = "dummy";
+	parts[4].offset = offset;
+	parts[4].size = 0;
+
+	parts[5].name = "art";
+	parts[5].offset = art_offset;
+	parts[5].size = TPLINK_ART_LEN;
+
 #else
 	parts[3].name = "art";
 	parts[3].offset = art_offset;
